### DISCUSSION TITLE
What are your biggest complaints about TShock?

### DISCUSSION THREAD

**Discussion by hakusaro** [Pryaxis/TShock#1734] [General]
Normally we ask for bug reports with reproduction steps. This time, we want the opposite: anything that bothers you. No need for details. Similar to what [Ruby on Rails' May of WTFs](https://discuss.rubyonrails.org/c/a-may-of-wtfs/8) is doing, what are the big issues that bother you that aren't necessarily bugs?

Asked in Slack, we got inconsistent errors as feedback. What else? What are the random sore spots in TShock that bother you? This can be commands, APIs, parts of the system, etc? What are the things we should definitely work on that aren't really bugs (or if they are bugs, maybe just tell us what they are so we know what we need to prioritize)?

**Comment by hakusaro** [comment]
From Discord: it's really unclear whether or not command line flags that override things are taking effect or not. If you set a port or max players on the command line or in the config file, we output different messages noting the override and making it confusing which one is happening when: https://discordapp.com/channels/479657350043664384/479657350043664386/709679152705568829

**Comment by hakusaro** [comment]
From Discord: the group system is really unclear in terms of API access and how to set groups:

> not sure why the error message is mentioning SQL however
> okay so the reason why this problem exists is cause Group exists in two locations TSPlayer and Account when using this method to update groups  TShock.UserAccounts.SetUserGroup() even thou it updates the database with the group
> and it updates TSPlayer.Group
> TSPlayer.Account.Group is not updated (or atleast its reference isn't)

https://discordapp.com/channels/479657350043664384/479657350043664386/709671417049776180

**Comment by moisterrific** [comment]
I believe I brought this up a long time ago before in the Pryaxis Discord, there is no standardized system for plugin devs to follow on how their plugin files work alongside TShock. 

From my experience hosting a server (mobile, though I would assume this same annoyance is on PC too) and when updating/porting outdated plugins - there is no standard solution for where a config or database file should be located within TShock’s directory, what usually ended up happening for me was the same folder that contained TShock’s very own config files got filled up with other plugins (config, database, etc) which looked very messy. 

Some plugin devs chose to do it more elegantly where they contain everything except for the .dll in a folder. Others just stuff everything in the same place TShock config go. 

I think it would be better if there was a standardized way devs should follow to make the directory look more organized and easier to find things.

**Reply by hakusaro** [reply]
Would it be better if we called the `tshock` folder something like `tshock_only_keep_out!!` or if we just published guidelines on what people should do? I get the feeling telling people what to do wouldn't really work. Maybe the API could return a plugin storage folder as a resource for devs to call?

**Reply by bartico6** [reply]
Yes, expose "data folder" from base plugin class.

This would lead to all plugins storing their data in a subfolder of same name as plugin, leading to much cleaner plugin folders.

It would look something like this, which is much more manageable than what tShock has now.
![Minecraft's Plugin Folder](https://i.quake.tf/1589275863-15702277.png)
For reference, that's what their base plugin exposes in order to enable this:
https://hub.spigotmc.org/javadocs/spigot/org/bukkit/plugin/java/JavaPlugin.html#getDataFolder--

**Comment by Kojirremer** [comment]
The first thing that comes to mind, hmm... I'd say the fact that when a user registers with a short number as their username, if you try to target said user with commands (for instance, /kick), TShock will instead use the command on a connected player that happens to be using the matching IX ID number.

It's not that bad, but it has been a bother a few times with unsavory individuals.
